{"metadata": {"language_info": {"mimetype": "text/x-python", "name": "python", "version": "3.6.3", "file_extension": ".py", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python"}, "kernelspec": {"display_name": "Python 3", "name": "python3", "language": "python"}}, "cells": [{"source": ["# Introduction\n", "--- \n", "\n", "I have a background in bioinformatics and this dataset was one of my favorites in grad school. For one of my projects, I analyzed the data in R but I have since adopted Python as my language of chocie. The aim of this kernel was to refresh myself and implement machine learning algorithms in Python. I accomplished what I wanted but I keep coming back to try new things.  The current state of this kernel is a comparison of GridSearchCV and RandomizedSearchCV for tuning hyperparameters with 10-fold cross-validation and eventually I would like to implement BayesianSearch for parameters.\n", "\n", "A heafty chunk of this kernel is just processing the data, so feel free to fork the kernel and use that part. The last half of the kernel is the comparison of the two parameter search methods. All in all, I found that both of these methods led to overfitting of the training data -- none of the models generalized to test data at all. That's the gist of what I learned but there is a little more discussion about it at the end of the kernel.\n", "\n", "Enjoy!\n", "<br><br>\n", "\n", "# About the original study\n", "I need to re-read the study, but from what I recall, these gene expression values come from cancer  patients with either acute lymphocytic leukemia (ALL) or acute myeloid leukemia (AML). In the original study, this datset was used to classify which type of cancer each patient had based on measurements of their gene expressions. The study was published in 1999 and was the first(?) to show that cancer types can be determined based on gene expressions alone. \n", "\n", "# About the data\n", "- Each row represents a different gene\n", "- Columns 1 and 2 are descriptions about that gene \n", "- Each numbered column is a patient\n", "- Each patient has 7129 gene expression values - i.e each patient has one value for each gene \n", "- The **training** data contain gene expression values for patients 1 through 38\n", "- The **test** data contain gene expression values for patients 39 through 72\n", "<br><br>\n", "\n", "\n", "# To do list:\n", "---\n", "### General cleanup\n", "- Write more functions for redundant testing (maybe, long term wish list)\n", "\n", "### Dimentionality reduction\n", "- Kanavanand has a great PCA analysis [here](https://www.kaggle.com/kanav0183/pca-analysis-for-geneclassification)\n", "\n", "### Tuning Hyperparamters:\n", "- Compare GridSearchCV to Bayesian Optimzation (I heard it's more efficient)\n", "    \n", "<br><br>"], "metadata": {"_cell_guid": "91c823f8-97b6-4652-b621-7d5580d1e30a", "_uuid": "e7d47224a60b48c834d8df8ff97c97e28c9c64aa"}, "cell_type": "markdown"}, {"source": ["# Data processing steps\n", "\n", "1. Remove columns that contain \"Call\" data, I'm not sure what they are but they doesn't seem useful\n", "2. Transpose the dataframe so that each row is a patient and each column is a gene\n", "3. Remove gene description and set the gene accession numbers as the column headers\n", "4. Merge the data (expression values) with the class labels (patient numbers)"], "metadata": {"_cell_guid": "2491bdb1-ee55-4545-b41a-c427e1dc7718", "_uuid": "b9b5f2a9ce59d7086e508751553e7d8392c390ad"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "e96760a5-4ff1-4261-a2c8-1a9e2e1bd880", "_kg_hide-input": false, "_execution_state": "idle", "collapsed": true, "_uuid": "8a1b2cf36761d30c89f72c1c2be37034b2733c59"}, "execution_count": null, "source": ["import itertools\n", "import matplotlib.pyplot as plt \n", "import numpy as np \n", "import pandas as pd\n", "import scipy"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "6a0cfafc-8afa-41c6-ba7a-fa61596bab1f", "_kg_hide-input": false, "_execution_state": "idle", "collapsed": true, "_uuid": "05345f350192b24c9ea7dac7e36e673f5f6a9db8"}, "execution_count": null, "source": ["testfile='../input/data_set_ALL_AML_independent.csv'\n", "trainfile='../input/data_set_ALL_AML_train.csv'\n", "patient_cancer='../input/actual.csv'\n", "\n", "train = pd.read_csv(trainfile)\n", "test = pd.read_csv(testfile)\n", "patient_cancer = pd.read_csv(patient_cancer)"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "c52961c6-9e65-45b5-8967-b4caf1978391", "_uuid": "689db3f563881b7c6f2441d1696cde299b200a40", "_execution_state": "idle", "collapsed": true}, "execution_count": null, "source": ["train.head()"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "f07b2efe-d87c-42b9-8d7d-f774cd3cb78b", "_kg_hide-input": false, "_execution_state": "idle", "collapsed": true, "_uuid": "a0f8f43756a6947faa1d4aed16035ac944c1c397"}, "execution_count": null, "source": ["# Remove \"call\" columns from training a test dataframes\n", "train_keepers = [col for col in train.columns if \"call\" not in col]\n", "test_keepers = [col for col in test.columns if \"call\" not in col]\n", "\n", "train = train[train_keepers]\n", "test = test[test_keepers]\n"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "577a80e1-8ee1-464b-973a-698189ecc9e6", "_uuid": "8493380de0d84b7b5dd2d7906122c7876466f020", "_execution_state": "idle", "collapsed": true}, "execution_count": null, "source": ["train.head()"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "b123fe25-65c4-42b9-8da4-99d93f3b11fc", "_uuid": "f444169669f997a125418a3cb900ec421396dd94", "collapsed": true}, "execution_count": null, "source": ["# Transpose the columns and rows so that genes become features and rows become observations\n", "train = train.T\n", "test = test.T\n", "train.head()"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "15eefd6d-4962-4277-814a-95bfa7a06dc7", "_uuid": "d7a9b8fc01f3e0022f55394e87b651a962b99ab5", "collapsed": true}, "execution_count": null, "source": ["# Clean up the column names for training data\n", "train.columns = train.iloc[1]\n", "train = train.drop([\"Gene Description\", \"Gene Accession Number\"]).apply(pd.to_numeric)\n", "\n", "# Clean up the column names for training data\n", "test.columns = test.iloc[1]\n", "test = test.drop([\"Gene Description\", \"Gene Accession Number\"]).apply(pd.to_numeric)\n", "\n", "train.head()"], "cell_type": "code"}, {"source": ["### Combine the data (gene expression) with class labels (patient numbers)"], "metadata": {"_cell_guid": "fa6672f2-c5a4-43cb-8077-6117c333cf3f", "_uuid": "66553b5ac208a39eb18fa9fcbfb2d1258484bab6"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "16fbe1c5-0952-4eb5-81b4-d452d0430c5e", "_kg_hide-input": false, "_uuid": "15906fae2bdd07f787d2b48a175ccc0859be1ded", "collapsed": true}, "execution_count": null, "source": ["# Reset the index. The indexes of two dataframes need to be the same before you combine them\n", "train = train.reset_index(drop=True)\n", "\n", "# Subset the first 38 patient's cancer types\n", "pc_train = patient_cancer[patient_cancer.patient <= 38].reset_index(drop=True)\n", "\n", "# Combine dataframes for first 38 patients: Patient number + cancer type + gene expression values\n", "train = pd.concat([pc_train,train], axis=1)\n", "\n", "\n", "# Handle the test data for patients 38 through 72\n", "# Clean up the index\n", "test = test.reset_index(drop=True)\n", "\n", "# Subset the last patient's cancer types to test\n", "pc_test = patient_cancer[patient_cancer.patient > 38].reset_index(drop=True)\n", "\n", "# Combine dataframes for last patients: Patient number + cancer type + gene expression values\n", "test = pd.concat([pc_test,test], axis=1)\n"], "cell_type": "code"}, {"source": ["# EDA\n", "---\n", "\n", "There's a bunch of data, so to speed things up I'm only using a small sample of the training data for the EDA.\n"], "metadata": {"_cell_guid": "44cd4399-9cd3-49b8-aa3d-515ad2c04756", "_uuid": "8075fadf78a3b76b67b40f31a647e9ad42c5bbec", "collapsed": true}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "06a40eb4-0b38-48f5-83a6-e8a9a4fdd9c9", "_uuid": "38b9fb6750ddc2f90202977097eef7a3ef0e6a5f", "collapsed": true}, "execution_count": null, "source": ["sample = train.iloc[:,2:].sample(n=100, axis=1)\n", "sample[\"cancer\"] = train.cancer\n", "sample.describe().round()"], "cell_type": "code"}, {"source": ["# To standardize or not to standardize\n", "---\n", "\n", "This is for a visual reference on how data changes after scaling. And for the record, I use the words standardize and scale interchangably. I think it's technically \"standardizing\", but scikit-learn calls it scaling. \n", "\n", "Standardize  = For each value, subtract the mean and scale to unit variance\n", "\n", "\n"], "metadata": {"_cell_guid": "8cb4370c-47f6-49b2-bcf2-8edbae6534b3", "_uuid": "a68ecc8483aa64dac93c2417d6a95c2008d90d26"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "dc7e9d78-5f66-40c4-8dc0-a8abe193a7be", "_uuid": "f60226125761e3b856a8e3868f702534b4eac119", "collapsed": true}, "execution_count": null, "source": ["from sklearn import preprocessing"], "cell_type": "code"}, {"source": ["### Distribution of the random sample before standardizing\n", "---"], "metadata": {"_cell_guid": "302dd036-e213-4a52-812f-28060c151456", "_uuid": "43d84c88bfe44e3f63500a83ff11989ab575fff6"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "a11f2f62-7166-412f-adc5-24cdb7566259", "_uuid": "a9064b7a0c313ecdb224f7e91088e6a55b910695", "collapsed": true}, "execution_count": null, "source": ["sample = sample.drop(\"cancer\", axis=1)\n", "sample.plot(kind=\"hist\", legend=None, bins=20, color='k')\n", "sample.plot(kind=\"kde\", legend=None);"], "cell_type": "code"}, {"source": ["Depending on the random sample and the histogram-bin-kung-fu, the data usually has a long skinny tail to the right. This KDE plot shows the distribution of indivudal features, but it's not very helpful here. This will change after standardizing the data. "], "metadata": {"_cell_guid": "4d7e6ddb-b492-41df-86c4-e8bc5e72f983", "_uuid": "c9958f8034ea1ee1a780a7efd552a308a174c358"}, "cell_type": "markdown"}, {"source": ["### Distribution of the random sample after standardizing\n", "---\n"], "metadata": {"_cell_guid": "1d98c27a-6ad0-4d98-945f-eff548e03e14", "_uuid": "ed323f625db8c06509c01098f2ed17f0917e2b90"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "5394ce94-d4c1-40aa-be25-258c5234e7e9", "_uuid": "2be76f45c36fcd1b5172d49930f6cd1593478be1", "collapsed": true}, "execution_count": null, "source": ["sample_scaled = pd.DataFrame(preprocessing.scale(sample))\n", "sample_scaled.plot(kind=\"hist\", normed=True, legend=None, bins=10, color='k')\n", "sample_scaled.plot(kind=\"kde\", legend=None);"], "cell_type": "code"}, {"source": ["There's quite a difference after standardizing the features. The KDE plot is much more useful in showing the individual distributions too. This is the result of subtracting the mean. Subtracting the mean from each feature centers them on zero. Neat!"], "metadata": {"_cell_guid": "68e94ce9-f94f-4907-a818-f9dc9621d512", "_uuid": "9de081773528e24250294797d67cd90295a086fa"}, "cell_type": "markdown"}, {"source": ["# Process the full set\n", "---\n", "\n", "\n"], "metadata": {"_cell_guid": "5fca5861-3762-4f27-b18a-564098709935", "_uuid": "a118162f00c884a29e9159463e7e6dc53a37ee86"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "6747e63f-4aca-43d1-8ae9-3dae56519d8f", "_uuid": "4356fd7bba7318cf983b777cf7fbddd742b5e07c", "collapsed": true}, "execution_count": null, "source": ["# StandardScaler to remove mean and scale to unit variance\n", "from sklearn.preprocessing import StandardScaler"], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "b26459eb-e2bb-4a06-bd71-60579d021e2f", "_uuid": "dabc60230623be1c1723862cc264316d8d43a943", "collapsed": true}, "execution_count": null, "source": ["scaler = StandardScaler().fit(train.iloc[:,2:])\n", "scaled_train = scaler.transform(train.iloc[:,2:])\n", "scaled_test = scaler.transform(test.iloc[:,2:])\n", "\n", "x_train = train.iloc[:,2:]\n", "y_train = train.iloc[:,1]\n", "x_test = test.iloc[:,2:]\n", "y_test = test.iloc[:,1]"], "cell_type": "code"}, {"source": ["\n", "# Classifiers\n", "---"], "metadata": {"_cell_guid": "2194e6ad-a24c-4835-a64d-a2354ae4e534", "_uuid": "f5244e36655b163a56b8e87f9a01e0f00d72af6d"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "e679a209-d341-4be3-8cb4-96a331307208", "_uuid": "2e7fa7c6bc6517db7c42df9bec581ee50be14fb8", "collapsed": true}, "execution_count": null, "source": ["# Grid Search for tuning parameters\n", "from sklearn.model_selection import GridSearchCV\n", "# RandomizedSearch for tuning (possibly faster than GridSearch)\n", "from sklearn.model_selection import RandomizedSearchCV\n", "# Bayessian optimization supposedly faster than GridSearch\n", "from bayes_opt import BayesianOptimization\n", "\n", "# Metrics\n", "from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, log_loss\n", "\n", "## Models\n", "from sklearn.linear_model import LogisticRegression\n", "from sklearn.svm import SVC\n", "from sklearn.neighbors import KNeighborsClassifier\n", "from sklearn.tree import DecisionTreeClassifier"], "cell_type": "code"}, {"source": ["# Helper functions"], "metadata": {"_cell_guid": "9efcd6c0-34ed-4b9b-bb78-86920ab14a34", "_uuid": "58ac80082fa2f28486225743dd9abbd69585eae0"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "5fe5062a-33ce-4504-8755-ce0820cf961f", "_uuid": "13cdd8de98183c59596132d9a7e50da910758b62", "collapsed": true}, "execution_count": null, "source": ["# CHERCHEZ FOR PARAMETERS\n", "def cherchez(estimator, param_grid, search):\n", "    \"\"\"\n", "    This is a helper function for tuning hyperparameters using teh two search methods.\n", "    Methods must be GridSearchCV or RandomizedSearchCV.\n", "    Inputs:\n", "        estimator: Logistic regression, SVM, KNN, etc\n", "        param_grid: Range of parameters to search\n", "        search: Grid search or Randomized search\n", "    Output:\n", "        Returns the estimator instance, clf\n", "    \n", "    \"\"\"   \n", "    try:\n", "        if search == \"grid\":\n", "            clf = GridSearchCV(\n", "                estimator=estimator, \n", "                param_grid=param_grid, \n", "                scoring=None,\n", "                n_jobs=-1, \n", "                cv=10, \n", "                verbose=0,\n", "                return_train_score=True\n", "            )\n", "        elif search == \"random\":           \n", "            clf = RandomizedSearchCV(\n", "                estimator=estimator,\n", "                param_distributions=param_grid,\n", "                n_iter=10,\n", "                n_jobs=-1,\n", "                cv=10,\n", "                verbose=0,\n", "                random_state=1,\n", "                return_train_score=True\n", "            )\n", "    except:\n", "        print('Search argument has to be \"grid\" or \"random\"')\n", "        sys.exit(0)\n", "        \n", "    # Fit the model\n", "    clf.fit(X=scaled_train, y=y_train)\n", "    \n", "    return clf   "], "cell_type": "code"}, {"outputs": [], "metadata": {"_cell_guid": "a27e8eec-79d4-45ee-a03c-9df9a8baab04", "_uuid": "e229a49752781655d6d5decd20e170b2feb0b9c8", "collapsed": true}, "execution_count": null, "source": ["# Function for plotting the confusion matrices\n", "def plot_confusion_matrix(cm, title=\"Confusion Matrix\"):\n", "    \"\"\"\n", "    Plots the confusion matrix. Modified verison from \n", "    http://scikit-learn.org/stable/auto_examples/model_selection/plot_confusion_matrix.html\n", "    Inputs: \n", "        cm: confusion matrix\n", "        title: Title of plot\n", "    \"\"\"\n", "    classes=[\"AML\", \"ALL\"]    \n", "    plt.imshow(cm, interpolation='nearest', cmap=plt.cm.bone)\n", "    plt.title(title)\n", "    tick_marks = np.arange(len(classes))\n", "    plt.xticks(tick_marks, classes, rotation=0)\n", "    plt.yticks(tick_marks, classes)\n", "    plt.ylabel('Actual')\n", "    plt.xlabel('Predicted')\n", "    thresh = cm.mean()\n", "    for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):\n", "        plt.text(j, i, format(cm[i, j]), \n", "                 horizontalalignment=\"center\",\n", "                 color=\"white\" if cm[i, j] < thresh else \"black\")    "], "cell_type": "code"}, {"source": ["# Models being tested\n", "1. Logisitc Regresison\n", "  - Using Grid search and Randomized search for tuning hyperparameters\n", "2. C-Support Vector Classification (SVM)\n", "  - Using Grid search and Randomized search for tuning hyperparameters\n", "3. K-Nearest Neighbors Classifier\n", "  - Using Grid search and Randomized search for tuning hyperparameters\n", "4. Decision Tree Classifier\n", "  - Using only Grid search"], "metadata": {"_cell_guid": "f247ca48-0b70-481e-acfb-628d85d088db", "_uuid": "783035f0903958d75e3ef8059070d86c0d7c0c0e"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "1e4e148e-1c96-4f5b-bd6d-f85c3f3c35d5", "scrolled": false, "_uuid": "4cb5bedd07807e9e98e92e1d735f05a6ffd2d9cc", "collapsed": true}, "execution_count": null, "source": ["# Logistic Regression\n", "# Paramaters\n", "logreg_params = {} \n", "logreg_params[\"C\"] =  [0.01, 0.1, 10, 100]\n", "logreg_params[\"fit_intercept\"] =  [True, False]\n", "logreg_params[\"warm_start\"] = [True,False]\n", "logreg_params[\"random_state\"] = [1]\n", "\n", "lr_dist = {}\n", "lr_dist[\"C\"] = scipy.stats.expon(scale=.01)\n", "lr_dist[\"fit_intercept\"] =  [True, False]\n", "lr_dist[\"warm_start\"] = [True,False]\n", "lr_dist[\"random_state\"] = [1]\n", "\n", "logregression_grid = cherchez(LogisticRegression(), logreg_params, search=\"grid\")\n", "acc = accuracy_score(y_true=y_test, y_pred=logregression_grid.predict(scaled_test))\n", "cfmatrix_grid = confusion_matrix(y_true=y_test, y_pred=logregression_grid.predict(scaled_test))\n", "print(\"**Grid search results**\")\n", "print(\"Best training accuracy:\\t\", logregression_grid.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "logregression_random = cherchez(LogisticRegression(), lr_dist, search=\"random\")\n", "acc = accuracy_score(y_true=y_test, y_pred=logregression_random.predict(scaled_test))\n", "cfmatrix_rand = confusion_matrix(y_true=y_test, y_pred=logregression_random.predict(scaled_test))\n", "print(\"**Random search results**\")\n", "print(\"Best training accuracy:\\t\", logregression_random.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "plt.subplots(1,2)\n", "plt.subplots_adjust(left=-0.5, bottom=None, right=None, top=None, wspace=0.5, hspace=None)\n", "plot_confusion_matrix(cfmatrix_rand, title=\"Random Search Confusion Matrix\")\n", "plt.subplot(121)\n", "plot_confusion_matrix(cfmatrix_grid, title=\"Grid Search Confusion Matrix\")"], "cell_type": "code"}, {"source": ["Discussion:<br>\n", "In this case, Logistic Regression wasn't very good at classifying between AML and ALL. In both Grid and Random searches, the differences between the training and test scores show that the model is overfitting to the training data. In fact, the random search results actually overfit more than the grid search which is not what I expected.  Neither search type improved actual prediction results. Comparing the accuracy in confusion matrices the true positives (top left and bottm right corners) are almost the same for each type of search.\n", "\n", "Conclusion: <br>\n", "A randomized parameter search resulted in higher variance. Possible caveats are that the dataset is rather small and I really haven't tried to choose proper parameters to tune so there may be room for a *little* improvement. \n", "<br><br>"], "metadata": {"_cell_guid": "2ed43540-c804-4a67-81e1-66173dcfca35", "_uuid": "ad472191395f53bb75e0cd4d389422b8d77e301d"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "9f8f1500-bf1a-4118-bd83-3c7b98263f9f", "_uuid": "35c0a28dffddca04064e270e9494a6bf430e9e98", "collapsed": true}, "execution_count": null, "source": ["# SVM\n", "svm_param = {\n", "    \"C\": [.01, .1, 1, 5, 10, 100],\n", "    \"gamma\": [0, .01, .1, 1, 5, 10, 100],\n", "    \"kernel\": [\"rbf\"],\n", "    \"random_state\": [1]\n", "}\n", "\n", "svm_dist = {\n", "    \"C\": scipy.stats.expon(scale=.01),\n", "    \"gamma\": scipy.stats.expon(scale=.01),\n", "    \"kernel\": [\"rbf\"],\n", "    \"random_state\": [1]\n", "}\n", "\n", "svm_grid = cherchez(SVC(), svm_param, \"grid\")\n", "acc = accuracy_score(y_true=y_test, y_pred=svm_grid.predict(scaled_test))\n", "cfmatrix_grid = confusion_matrix(y_true=y_test, y_pred=svm_grid.predict(scaled_test))\n", "print(\"**Grid search results**\")\n", "print(\"Best training accuracy:\\t\", svm_grid.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "svm_random = cherchez(SVC(), svm_dist, \"random\")\n", "acc = accuracy_score(y_true=y_test, y_pred=svm_random.predict(scaled_test))\n", "cfmatrix_rand = confusion_matrix(y_true=y_test, y_pred=svm_random.predict(scaled_test))\n", "print(\"**Random search results**\")\n", "print(\"Best training accuracy:\\t\", svm_random.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "plt.subplots(1,2)\n", "plt.subplots_adjust(left=-0.5, bottom=None, right=None, top=None, wspace=0.5, hspace=None)\n", "plot_confusion_matrix(cfmatrix_rand, title=\"Random Search Confusion Matrix\")\n", "plt.subplot(121)\n", "plot_confusion_matrix(cfmatrix_grid, title=\"Grid Search Confusion Matrix\")"], "cell_type": "code"}, {"source": ["Discussion:<br>\n", "Well, this example goes to show that if you just predict that every patient has AML, you'll be correct more often than wrong. The model predicted that all 34 patients in the test data had AML and was absolutely right for 20 of them but absolutely wrong for the other 14 patients. \n", "\n", "Conclusion<br>\n", "I wouldn't use this particular model to classify cancer patients.\n", "<br><br>"], "metadata": {"_cell_guid": "051f68f9-46f2-4147-9d0e-22318dee36be", "_uuid": "012fe8fcb29408ba4db9ce4a3ea858644cfc72d2"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "34bf611b-063d-462f-a238-58d6c7c1e3d8", "scrolled": false, "_uuid": "3d16144e47b7a79d22e00608439e2ef87cdb676d", "collapsed": true}, "execution_count": null, "source": ["# KNN\n", "knn_param = {\n", "    \"n_neighbors\": [i for i in range(1,30,5)],\n", "    \"weights\": [\"uniform\", \"distance\"],\n", "    \"algorithm\": [\"ball_tree\", \"kd_tree\", \"brute\"],\n", "    \"leaf_size\": [1, 10, 30],\n", "    \"p\": [1,2]\n", "}\n", "\n", "knn_dist = {\n", "    \"n_neighbors\": scipy.stats.randint(1,33),\n", "    \"weights\": [\"uniform\", \"distance\"],\n", "    \"algorithm\": [\"ball_tree\", \"kd_tree\", \"brute\"],\n", "    \"leaf_size\": scipy.stats.randint(1,1000),\n", "    \"p\": [1,2]\n", "}\n", "\n", "knn_grid = cherchez(KNeighborsClassifier(), knn_param, \"grid\")\n", "acc = accuracy_score(y_true=y_test, y_pred=knn_grid.predict(scaled_test))\n", "cfmatrix_grid = confusion_matrix(y_true=y_test, y_pred=svm_grid.predict(scaled_test))\n", "print(\"**Grid search results**\")\n", "print(\"Best training accuracy:\\t\", knn_grid.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "knn_random = cherchez(KNeighborsClassifier(), knn_dist, \"random\")\n", "acc = accuracy_score(y_true=y_test, y_pred=knn_random.predict(scaled_test))\n", "cfmatrix_rand = confusion_matrix(y_true=y_test, y_pred=knn_random.predict(scaled_test))\n", "print(\"**Random search results**\")\n", "print(\"Best training accuracy:\\t\", knn_random.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "plt.subplots(1,2)\n", "plt.subplots_adjust(left=-0.5, bottom=None, right=None, top=None, wspace=0.5, hspace=None)\n", "plot_confusion_matrix(cfmatrix_rand, title=\"Random Search Confusion Matrix\")\n", "plt.subplot(121)\n", "plot_confusion_matrix(cfmatrix_grid, title=\"Grid Search Confusion Matrix\")"], "cell_type": "code"}, {"source": ["Discusison:<br>\n", "As with the other models, an automated search for hyperparameters with KNN just overfits the training data and doesn't generalize to the test data. The one difference here is that a randomized search produced a model which predicted that at least some of the patients had ALL rather than predicting that all patients have AML.\n", "\n", "Conclusion:<br>\n", "My main problem is that the GridSearch and RandomSearch use the model that fits the training data the best without taking into account how well it will generalizes to the test data. "], "metadata": {"_cell_guid": "27c8510a-13e4-435c-8e6b-d0efa8d51042", "_uuid": "3e32bd77b448b84835ac7d0ee72c9e47879be15d"}, "cell_type": "markdown"}, {"outputs": [], "metadata": {"_cell_guid": "98378072-dc27-435e-9d9c-5059e46c32fa", "_uuid": "06948cdc2f306a18894678a09184f4f6f4ac70ca", "collapsed": true}, "execution_count": null, "source": ["# Decision tree classifier\n", "dtc_param = {\n", "    \"max_depth\": [None],\n", "    \"min_samples_split\": [2],\n", "    \"min_samples_leaf\": [1],\n", "    \"min_weight_fraction_leaf\": [0.],\n", "    \"max_features\": [None],\n", "    \"random_state\": [4],\n", "    \"max_leaf_nodes\": [None], # None = infinity or int\n", "    \"presort\": [True, False]\n", "}\n", "\n", "\n", "dtc_grid = cherchez(DecisionTreeClassifier(), dtc_param, \"grid\")\n", "acc = accuracy_score(y_true=y_test, y_pred=dtc_grid.predict(scaled_test))\n", "cfmatrix_grid = confusion_matrix(y_true=y_test, y_pred=dtc_grid.predict(scaled_test))\n", "print(\"**Grid search results**\")\n", "print(\"Best training accuracy:\\t\", dtc_grid.best_score_)\n", "print(\"Test accuracy:\\t\", acc)\n", "\n", "plot_confusion_matrix(cfmatrix_grid, title=\"Decision Tree Confusion Matrix\")"], "cell_type": "code"}, {"source": ["Discussion:<br>\n", "This model accurately predicted more ALL patients than any of the other models and had the lowest false positive rate, but frankly that's just painting it in a positive light and I'm not even bothering to calculate it. The model still has poor accuracy overall, doesn't generalize to the test data, and has a huge variance problem\n", "\n", "Conclusion:<br>\n", "This experiment makes a good case for splitting your data into train, validation, and test sets. It would be nice if GridSearch and RandomSearch could refit the best predictors that minimize the variance between the accuracy of train and validation sets. As it is, overfitting the training data has been a problem for each of these models. It's possible that I could investigate which parameters would be better to tune and it's possible that there isn't enough data, and it's also possible that this isn't a good classification problem. Maybe I need to approach this as a clustering problem. What do you think?"], "metadata": {"_cell_guid": "03d0cc30-e260-4d48-bd31-c90e5fe4db45", "_uuid": "2b611db83ae3242bc603c53abf28aaba848be509"}, "cell_type": "markdown"}, {"source": ["# Sources:\n", "\n", "Golub et al: https://www.ncbi.nlm.nih.gov/pubmed/10521349\n", "\n", "Bayesian optimization: https://arxiv.org/pdf/1012.2599v1.pdf\n"], "metadata": {"_cell_guid": "f1ee2958-b829-4039-b731-13acf10d1744", "_uuid": "9e254dfc3ecd04d657d6f88f6d5a81320d219976"}, "cell_type": "markdown"}], "nbformat": 4, "nbformat_minor": 1}